### PR TITLE
Updates 'grovetccfg' to to write a grove.cfg from the grove servers profile

### LINKED ISSUE #2607: grovetccfg does not update the grove.cfg file
grovetccfg is the grove equivalent of the ort script.  currently it only reads ip_allow parameters from a grove servers profile and only build remap rules.  It does not manage the grove.cfg or the settings therein.  PR #2606 address's this issue

### PR DESCRIPTION
**Pull request by jrushford**
Modified grovetccfg to read a grove server profile from the traffic_ops and write a grove.cfg from that profile id there have been changes. Fixes #2607 and depends on #2517 and #2523. Both #2517 and #2523 are merged onto master.

### CODE REVIEW

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/trafficcontrol-PR/2161/
Test FAILed.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/trafficcontrol-PR/2165/
Test PASSed.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/trafficcontrol-PR/2166/
Test PASSed.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/trafficcontrol-PR/2167/
Test PASSed.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/trafficcontrol-PR/2170/
Test PASSed.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/trafficcontrol-PR/2171/
Test PASSed.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/trafficcontrol-PR/2172/
Test PASSed.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/trafficcontrol-PR/2174/
Test PASSed.

File: grove/grovetccfg/grovetccfg.go [hunk 1]
```diff
@@ -259,20 +328,162 @@ func main() {
 	os.Exit(0)
 }
 
-func createRulesOldAPI(toc *to.Session, host string, certDir string) (remap.RemapRules, error) {
-	cachegroupsArr, err := toc.CacheGroups()
+func createGroveCfg(toc *to.Session, server tc.Server) (bool, config.Config, error) {
+	var new_cfg config.Config
+	var cur_cfg config.Config
+	var plugin_params = []string{}
```
Comment (rob05c) [inline grove/grovetccfg/grovetccfg.go]: Go local variables should be `camelCase` not `snake_case`

File: grove/grovetccfg/grovetccfg.go [hunk 2]
```diff
@@ -259,20 +328,162 @@ func main() {
...
+	} else { // updates are required, send the new config
+		fmt.Println(time.Now().Format(time.RFC3339Nano) + " Config updates are required to '" + GroveConfigPath + "'")
+		return true, new_cfg, nil
 	}
-	cachegroups := makeCachegroupsNameMap(cachegroupsArr)
+}
+
+func setConfigParameter(cfg *config.Config, name string, value string) error {
+	var err error
+	err = nil
```
Comment (rob05c) [inline grove/grovetccfg/grovetccfg.go]: This is unnecessary, Go zero-initializes all variables, so `err` is already `nil`.

File: grove/grovetccfg/grovetccfg.go [hunk 3]
```diff
@@ -259,20 +328,162 @@ func main() {
...
+	if name == "rfc_compliant" {
```
Comment (rob05c) [inline grove/grovetccfg/grovetccfg.go]: A `switch` would be cleaner here.
```
switch name {
case "rfc_compliant":
  cfg.RFCCompliant, err = strconv.ParseBool(value)
case "remap_rules_file":
  cfg.RemapRulesFile = value
...
}
return err
```

File: grove/grovetccfg/grovetccfg.go [hunk 4]
```diff
@@ -94,13 +99,13 @@ func CopyAndGzipFile(src, dst string) error {
 }
 
 // BackupFile copies the given file to a new file in the subdirectory "/remap_history", with the name suffixed by the current timestamp. Returns nil if the given file doesn't exist (nothing to back up).
-func BackupFile(path string) error {
+func BackupFile(path string, bkup_dir string) error {
```
Comment (rob05c) [inline grove/grovetccfg/grovetccfg.go]: `bkup_dir` should be camel case, and while Go favors brevity `bkup` is a bit extreme, I'd recommend `backupDir`

File: grove/grovetccfg/grovetccfg.go [hunk 5]
```diff
@@ -202,11 +207,75 @@ func main() {
...
 	}
 
+	// TODO remove this once converted to 1.3 API
+	// using the 1.2 API
+	var hostName string
+	var hostServer tc.Server
+	var hostProfile tc.Profile
+	var ok bool
+	var profiles map[string]tc.Profile
+	var servers map[string]tc.Server
```
Comment (rob05c) [inline grove/grovetccfg/grovetccfg.go]: If `host == nil`, this gets passed to `createRulesOldAPI` which will panic. This should check if `host == nil`, and print a good error message and `os.Exit(1)`.

That will also reduce indentation, and reduce variable pre-declarations too, which will make this block much nicer.

File: grove/grovetccfg/grovetccfg.go [hunk 6]
```diff
@@ -202,11 +207,75 @@ func main() {
...
+
+		hostProfile, ok = profiles[hostServer.Profile]
+		if !ok {
+			fmt.Println(time.Now().Format(time.RFC3339Nano) + " Error: profile '" + hostServer.Profile + "' not in Profiles\n")
+			os.Exit(1)
+		}
+	} // end of API 1.2 stuff
+
+	if hostProfile.Type == GroveProfileType {
+		update_required, cfg, err := createGroveCfg(toc, hostServer)
```
Comment (rob05c) [inline grove/grovetccfg/grovetccfg.go]: `update_required` should be `updateRequired`

File: grove/grovetccfg/grovetccfg.go [hunk 7]
```diff
@@ -259,20 +328,162 @@ func main() {
...
+	serverParameters, err := toc.Parameters(server.Profile)
 	if err != nil {
-		fmt.Println(time.Now().Format(time.RFC3339Nano) + " Error getting Traffic Ops Cachegroups: " + err.Error())
-		os.Exit(1)
+		fmt.Println(time.Now().Format(time.RFC3339Nano) + " Error getting Traffic Ops Parameters for host '" + server.HostName + "' profile '" + server.Profile + "': " + err.Error())
+		return false, cur_cfg, err
+	} else {
+		// load config parameters from the servers profile
+		for _, p := range serverParameters {
+			if p.ConfigFile == "grove.cfg" {
```
Comment (rob05c) [inline grove/grovetccfg/grovetccfg.go]: The config file is duplicated in `GroveConfigPath`. Can you change the variables to 
```
const GroveConfigFile = "grove.cfg"
const GroveConfigPath = "/etc/grove/" + GroveConfigFile
```

To get rid of the magic string?

File: grove/grovetccfg/grovetccfg.go [hunk 8]
```diff
@@ -259,20 +328,162 @@ func main() {
...
-func createRulesOldAPI(toc *to.Session, host string, certDir string) (remap.RemapRules, error) {
-	cachegroupsArr, err := toc.CacheGroups()
+func createGroveCfg(toc *to.Session, server tc.Server) (bool, config.Config, error) {
+	var new_cfg config.Config
+	var cur_cfg config.Config
+	var plugin_params = []string{}
+
+	// load the servers current config parameters.
+	if _, err := os.Stat(GroveConfigPath); err == nil {
+		cur_cfg, err = config.LoadConfig(GroveConfigPath)
```
Comment (rob05c) [inline grove/grovetccfg/grovetccfg.go]: `cur_cfg` should be `currentCfg` or `currCfg`

File: grove/grovetccfg/grovetccfg.go [hunk 9]
```diff
@@ -259,20 +328,162 @@ func main() {
...
+}
+
+func setConfigParameter(cfg *config.Config, name string, value string) error {
+	var err error
+	err = nil
+	if name == "rfc_compliant" {
+		cfg.RFCCompliant, err = strconv.ParseBool(value)
+		return err
+	}
+	if name == "port" {
```
Comment (rob05c) [inline grove/grovetccfg/grovetccfg.go]: All these strings duplicating the JSON tags aren't great, but unfortunately, I don't think Go has a better way to do this

File: grove/grovetccfg/grovetccfg.go [hunk 10]
```diff
@@ -50,6 +52,9 @@ const UserAgent = "grove-tc-cfg/" + Version
 const TrafficOpsTimeout = time.Second * 90
 const DefaultCertificateDir = "/etc/grove/ssl"
 const GroveConfigPath = "/etc/grove/grove.cfg"
+const ConfigHistory = "cfg_history/"
+const RemapHistory = "remap_history/"
+const GroveProfileType = "GROVE_PROFILE"
```
Comment (rob05c) [inline grove/grovetccfg/grovetccfg.go]: Would you mind moving this to `traffic_control/lib/go-tc/enum.go`? Profile types should be there, so other apps like the TO client can use them.

**Comment by rob05c** [comment]
Ok, I commented on all the formatting things I saw. I didn't do an actual review, because that will take longer, and I'll need to install and run it against a TO instance with the expected profile. I'll try to do that as soon as I can find the time.

Feel free to fix the things noted now, or wait until the full review, whatever you prefer.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/trafficcontrol-PR/2189/
Test FAILed.

**Comment by jrushford** [comment]
Hey Rob,

I’ve made the requested changes except for this one, I do not see that there is currently any profile enum in go-tc.  If necessary, I could add one I suppose.
I had several commits so, I’ve also squashed them.

If it helps, on ipcdn-cache-59.cdnlab.comcast.net, I have a cdn-in-a-box running with the current database from staging where I have added the GROVE_PROFILE to profile_type_values and  have a EDGE1_XCR_GROVE_0.1-9001 built.  I’ve been running this script on cache-59 for testing, you could use my hostname below or pick another and change it’s profile to use this one:

#!/bin/bash
./grovetccfg -ignore-update-flag -host cdn-ec-sfb-001-01 -touser jrushford -topass twelve -insecure -tourl https://localhost:6443 -pretty

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/trafficcontrol-PR/2191/
Test FAILed.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/trafficcontrol-PR/2194/
Test PASSed.

**Comment by jrushford** [comment]
@dewrich, @dangogh, @DylanVolz can i get a review of this pr.  It is pretty straight forward.  It writes a grove.cfg file for a server when a grove profile is used.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/trafficcontrol-PR/2352/
Test PASSed.

**Pull request merged**